### PR TITLE
Optim removal and leverage lr_scheduler from pytorch

### PR DESCRIPTION
**Pull request by JianyuZhan**
This PR addresses Issue https://github.com/OpenNMT/OpenNMT-py/issues/233. It also fixes the `train_from` broken logic,  now we don't need to load `Optimizer` from checkpoint, then setup a new one and throw away the loaded one. See Testing section below for details.

**Overview**

In this PR, we leverage `lr_scheduler` to do the learning rate update. Thus we could get rid of `Optim.py`. We have now 3 types of learning rate update mechanisms:

1.  Batch level update, it will be updated at batch border(thus several times in an epoch). I delegate to `lr_scheduler.LambdaLR` for this duty.

2. Epoch level, `start_decay_at` logic. I also delegate to  `lr_scheduler.LambdaLR` for this duty.

3. Epoch level, `perplexity not dreasing in this epoch` logic. I delegate to `lr_scheduler.ReduceLROnPlateau` for this duty.

Note, in each epoch, if both 2 and 3 are satisfied, only one lr update is performed,  I keep this behavior in the new code.

See `NMTModel.setup_lr_scheduler()` for this code logic. 

 `NMTModel.setup_optimizer()`  is for setting up the `Optimizer`.

The original training side interface is kept:

`optim.step()` => `model.optimizer_step()`
`optim.updateLearningRate(ppl, epoch)` => `model.update_learning_rate(ppl, epoch)`

The `Optim` class is deleted, so now `checkpoint['optim']` stores the `torch.optim.Optimizer` instead.

**Testing**

To test this code. I add printing lr update log in code(I left the print code, with an arg `verbose` default to be False, in case debugging is needed, just toggle it on), and analyze a `SRU` training and a transformer training log, in which I observed:

1.  batch level update was performed at each batch(for the demo test, it has 157 batches for the demo example, with default `batch_size` 64).

2.  `start_decay_at` logic started to halve the lr(default factor: 0.5) from epoch 8(default) onward.

3.  when the validity perplexity did not decrease in an epoch, the perplexity update logic was performed.

Besides, I fixed the `train_from` broken logic,  now we don't need to load `Optimizer` from checkpoint, then setup a new one and throw away the loaded one.  In a test case, I first performed a `-epochs 3` test, and then used its output to perform the second `-train_from xxx_e3.pt` run,  I observed that the lr is restored to the point before, together with the `step` count resumed, so the `noam` factor is calculated correctly. Also, it correctly performed `start_decay_at` logic in epoch 8. All in all, it just acted like a fresh training from beginning.

`test/rebuild_test_models.sh` is run to rebuild the test models since we've changed the `checkpoint['optim']`.

**Problem**

 Our versoin of `noam` is different from the one of `tensor2tensor`, which don't have the multiplicative factor `rnn_size ** (-0.5)`. I still keep this logic. But I am not sure if it is desired, I observed that lr is changing sort of like this with `noam`  while testing, it falls to 0.0 at first, then climbs up slowly, and after epoch 8 or ppl doesn't decrease in that epoch, it would be halved.

```
Intializing params.
* number of parameters: 52108320
('encoder: ', 14000500)
('decoder: ', 38107820)
()
Before batch update, lr: 1.000000
After batch update, lr: 0.000000
Before batch update, lr: 0.000000
After batch update, lr: 0.000000
Before batch update, lr: 0.000000
After batch update, lr: 0.000001
Before batch update, lr: 0.000001
After batch update, lr: 0.000001
Before batch update, lr: 0.000001
After batch update, lr: 0.000001
Before batch update, lr: 0.000001
After batch update, lr: 0.000001
Before batch update, lr: 0.000001
After batch update, lr: 0.000001
Before batch update, lr: 0.000001
After batch update, lr: 0.000001
Before batch update, lr: 0.000001
After batch update, lr: 0.000002
Before batch update, lr: 0.000002
After batch update, lr: 0.000002
Before batch update, lr: 0.000002
After batch update, lr: 0.000002
Before batch update, lr: 0.000002
After batch update, lr: 0.000002
Before batch update, lr: 0.000002
After batch update, lr: 0.000002
Before batch update, lr: 0.000002
.....

### CODE REVIEW

File: onmt/Loss.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def nmt_criterion(vocab_size, gpuid, pad_id):
     return crit
 
 
-class Statistics:
+class Statistics(object):
```
Comment (srush) [inline onmt/Loss.py]: Maybe move this to a new file called Log.py?  Possibly removing the score() method and just making a local function.

File: onmt/Models.py [hunk 2]
```diff
@@ -425,6 +429,117 @@ def forward(self, src, tgt, lengths, dec_state=None):
             attns = None
         return out, attns, dec_state
 
+    @property
```
Comment (srush) [inline onmt/Models.py]: No I disagree with this. None of this should be in Models.py optimization should be basically independent of models.

File: onmt/Models.py [hunk 3]
```diff
@@ -425,6 +429,117 @@ def forward(self, src, tgt, lengths, dec_state=None):
...
+    def lr(self):
+        return self.optimizer.param_groups[0]['lr']
+
+    @lr.setter
+    def lr(self, lr):
+        self.optimizer.param_groups[0]['lr'] = lr
+
+    def setup_optimizer(self, optimizer, initial_lr, max_grad_norm):
```
Comment (srush) [inline onmt/Models.py]: Think this should go in the ModelController file. Maybe should be just called Controller now.

File: onmt/Models.py [hunk 4]
```diff
@@ -425,6 +429,117 @@ def forward(self, src, tgt, lengths, dec_state=None):
...
+
+        self.lr_threshold_scheduler = lr_scheduler.LambdaLR(
+                self.optimizer, lr_lambda=threshold_lr, last_epoch=last_epoch)
+
+        # 3. Epoch level lr update: only when perplexity doesn't decrease.
+        self.lr_ppl_scheduler = lr_scheduler.ReduceLROnPlateau(
+                self.optimizer, 'min',
+                factor=lr_decay, patience=0, verbose=True)
+
+    def optimizer_step(self, verbose=False):
```
Comment (srush) [inline onmt/Models.py]: I think this should be directly in the training loop now.

File: onmt/Models.py [hunk 5]
```diff
@@ -425,6 +429,117 @@ def forward(self, src, tgt, lengths, dec_state=None):
...
+            last_lr_steps = self.optimizer.state['last_lr_steps']
+        else:
+            last_epoch = 0  # for epoch-level lr update.
+            last_lr_steps = 0  # for batch-level lr update.
+
+        # LambdaLR requires 'initial_lr' set if last_epoch is passed in.
+        self.optimizer.param_groups[0]['initial_lr'] = self.lr
+
+        # 1. Batch level lr update: using 'noam' method.
+        self.lr_batch_scheduler = None
```
Comment (srush) [inline onmt/Models.py]: Do these all need to be member variables? The aim of the PR was to have less state associated with Optimization.

File: train.py [hunk 6]
```diff
@@ -83,7 +83,7 @@ def eval(model, criterion, data, fields):
     return stats
 
 
-def train_model(model, train_data, valid_data, fields, optim):
+def train_model(model, train_data, valid_data, fields):
```
Comment (srush) [inline train.py]: Can we also move this to a onmt/Trainer.py file with anything else we might need for training and optimization. Would like to be able to train without using the command-line.

File: train.py [hunk 7]
```diff
@@ -183,28 +183,29 @@ def train_epoch(epoch):
...
+        model.update_learning_rate(valid_stats.ppl(), epoch)
 
-        model_state_dict = (model.module.state_dict() if len(opt.gpuid) > 1
-                            else model.state_dict())
-        model_state_dict = {k: v for k, v in model_state_dict.items()
-                            if 'generator' not in k}
-        generator_state_dict = (model.generator.module.state_dict()
-                                if len(opt.gpuid) > 1
-                                else model.generator.state_dict())
         #  (4) drop a checkpoint
```
Comment (srush) [inline train.py]: This part can be a function in onmt/Train.py "save_checkpoint"

**Comment by srush** [comment]
See equation (3) in the attention is all you need paper https://arxiv.org/pdf/1706.03762.pdf. It definitely has that factor. I think it is being incorporated in their code directly for each model. We could do this I guess, but I like that it looks like the paper.

**Comment by srush** [comment]
Finally, it would be great if there was a way to test learning rates. Could you factor it in such a way that you could write a unitest to check that the learning rate is acting like it should (without training).

File: onmt/Models.py (hunk 3 above)
Comment (JianyuZhan) [reply onmt/Models.py]: I agree that this should be moved out of `NMTModel`. Since we would have a `Trainer.py`,  I think it makes more sense to be part of the duty of `Trainer`.    The static property of a model is in ` ModelConstructor.py`,  and the dynamic ones falls in the train phase and translate phase, which are `Trainer.py` and `Translator.py` , respectively.   As for the `Statistics` class,  I agree that it should be moved out `Loss.py` too,  but not in a `Log.py`.  The name is too vague, and it would guide people to think that it is logging utility or something like that.  I would like to put it in the `Trainer.py` too,  acting as a helper class for `Trainer`.

Further more,  this makes the most of the train logic self-containing in this file. Ideally,  after that, 
`train.py` would be acting as just a driver around the `Trainer`,  I think this could finish the https://github.com/OpenNMT/OpenNMT-py/issues/239.  

Comments ?  I will stop to start a new PR for this work first, after that I will come back for this PR.

**Comment by JianyuZhan** [comment]
I agree that `optim` stuff should be moved out of `NMTModel`. Since we would have a `Trainer.py`,  I think it makes more sense to be part of the duty of `Trainer`.    The static property of a model is in ` ModelConstructor.py`,  and the dynamic ones falls in the train phase and translate phase, which are `Trainer.py` and `Translator.py` , respectively. 

As for the `Statistics` class,  I agree that it should be moved out `Loss.py` too,  but not in a `Log.py`.  The name is too vague, and it would guide people to think that it is logging utility or something like that.  I would like to put it in the `Trainer.py` too,  acting as a helper class for `Trainer`.

Further more,  this makes the most of the train logic self-containing in `Trainer.py`. Ideally,  after that, `train.py` would be acting as just a driver around the `Trainer`,  I think this could drive us a step towards solving https://github.com/OpenNMT/OpenNMT-py/issues/239, and users(`train.py` is one of them) would get a pytorch-like interface more or less like this:

```
import onmt.Trainer

... preprare model, train data, validate data, fields...

trainer = Trainer(....)

for epoch in range(100):
    trainer.train(...)
    trainer.validate(...)
    if needed:
         trainer.log()
    trainer.optim_step()
    if needed:
          trainer.drop_checkpoint()
```

Comments ?  I will stop to start a new PR for this work first, after that I will come back for this PR.

**Comment by srush** [comment]
That plan seems good to me.

**Comment by srush** [comment]
Let me know when this one is ready to go again.

**Comment by JianyuZhan** [comment]
I am working on to migrate the code to master branch and retest, I think tomorrow I could resend the PR.

**Comment by JianyuZhan** [comment]
Ah, I just updated the branch and pushed to my remote, then github found there are no differences between it and the master, so it deemed it is merged and closed it automatically.  I will reopen it after I push the new code tomorrow.

**Comment by JianyuZhan** [comment]
v1 -> v2:

-  The `noam` factor still use `rnn_size`,  so this could be merged, when #287 is merged, I will then fix it to use general model size.
-  The new work is mostly in a new file `onmt/ModelController.py`. A `ModelController` is in charge of model dynamics, currently basically the optimizer, lr updater,  the grad norm clip. And now `Trainer` doesn't accept an `optim`, instead, it accepts a `model_controller`, see: https://github.com/JianyuZhan/OpenNMT-py/blob/optim_removal/train.py#L136
-  Since `noam` updates the lr based on the initial_lr ( `new_lr = initial_lr * new_factor`),  thus it effectively ignores the other two update methods,  so in this PR,  I explicitly turn off them if `noam` is used.  The other logic is kept the same as before.
- To test the lr update logic,  a new unittest is added: `test/test_lr_update.py`.  The approach is like this: I borrowed the old update code logic from the old `Optim.py`  as the baseline updater,  and then run two parallel updaters (one new, one baseline updater) under same settings(same options, same ppl, etc), and assert the new lr is the same.  Several settings are used to test, including the train_from case.
-  Added a new command line option `-lr_update_verbose`(default `False`) for user to toggle on if user wants to observe the lr update log.

**Test**
- 13 unittest cases passed.
-  real case run (with `-train_from`) passed, the final accurary/ppl is quite close to the baseline version.

**Comment by srush** [comment]
I no longer understand the point of the PR. It seems like you ended up just replacing Optim with ModelController and they do the same thing? What structural benefit does this provide?

File: onmt/Trainer.py [hunk 8]
```diff
@@ -169,11 +168,21 @@ def validate(self):
...
+        """ Called at each epoch to update learning rate.
+        We only do it, model_controller takes care of how to do it.
+        """
+        self.model_controller.epoch_step(ppl, epoch)
+
+    def _batch_step(self):
+        """ Called at each batch to update parameters and lr.
+        We only do it, model_controller takes care of how to do it.
+        """
+        self.model_controller.lr_step_noam()
```
Comment (srush) [inline onmt/Trainer.py]: It is bad that this code has to know the details of model controller.

File: onmt/Trainer.py [hunk 9]
```diff
@@ -56,44 +56,43 @@ def output(self, epoch, batch, n_batches, start):
...
         experiment.add_scalar_value(prefix + "_tgtper",  self.n_words / t)
-        experiment.add_scalar_value(prefix + "_lr", optim.lr)
+        experiment.add_scalar_value(prefix + "_lr", lr)
 
 
 class Trainer(object):
-    def __init__(self, model, train_iter, valid_iter,
-                 train_loss, valid_loss, optim,
-                 trunc_size, shard_size):
+    def __init__(self, model, model_controller, train_iter, valid_iter,
```
Comment (srush) [inline onmt/Trainer.py]: My hope was that an torch.optim object could be passed here directly.

**Comment by JianyuZhan** [comment]
I admitted this look quite structurally the same as before,  I initially would like to make it thin or without such a wrapper at all.  The torch code separates `torch.optim.optimizer` and `torch.optim.lr_scheduler`,  and the latter accepts the former as an arg.  So it structurally doesn't support a monolithic concept like `Optim` that contains all these.  Moreover,  we provide 3 lr update mechanisms,  no a single `lr_scheduler` could do all duties, so this is why I use different lr_schedulers.  And we have norm_grad_clip too.  All these finally drive me to create a `model_controller`, which hide all things about the model dynamic control.   So if we don't inflate the model,  the `Trainer` then has to know all these details.

**Comment by srush** [comment]
Let's talk on gitter when you get back from vacation.
